spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/reactive/PubSubReactiveFactory.java

```diff
@@ -0,0 +1,168 @@
+	 *
+	 * @param subscriberOperations template for interacting with GCP Pub/Sub subscriber operations.
+	 */
+	public PubSubReactiveFactory(PubSubSubscriberOperations subscriberOperations) {
+		Assert.notNull(subscriberOperations, "subscriberOperations cannot be null.");
+		this.subscriberOperations = subscriberOperations;
+	}
+
+	/**
+	 * Create an infinite stream {@link Flux} of {@link AcknowledgeablePubsubMessage} objects.
+	 * <p>The {@link Flux} respects backpressure by using of Pub/Sub Synchronous Pull to retrieve
+	 * batches of up to the requested number of messages until the full demand is fulfilled
+	 * or subscription terminated.
+	 * <p>For unlimited demand, the underlying subscription will be polled at a regular interval,
+	 * requesting up to {@code Integer.MAX_VALUE} messages at each poll.
+	 * <p>For specific demand, as many messages as are available will be returned immediately,
+	 * with remaining demand being fulfilled in the future.
+	 * Pub/Sub timeout will cause a retry with the same demand.
+	 * @param subscriptionName subscription from which to retrieve messages.
+	 * @param pollingPeriod how frequently to poll the source subscription in case of unlimited demand.
+	 * @return infinite stream of {@link AcknowledgeablePubsubMessage} objects.
+	 */
+	public Flux<AcknowledgeablePubsubMessage> createPolledFlux(String subscriptionName, long pollingPeriod) {
+
+		return Flux.create(sink -> {
+			sink.onRequest((numRequested) -> {
+				if (numRequested == Long.MAX_VALUE) {
+					// unlimited demand
+					Disposable task = Schedulers.single().schedulePeriodically(
+							new PubSubNonBlockingUnlimitedDemandPullTask(subscriptionName, sink), 0, pollingPeriod, TimeUnit.MILLISECONDS);
+					sink.onCancel(task::dispose);
+				}
+				else {
+					Schedulers.single().schedule(new PubSubBlockingLimitedDemandPullTask(subscriptionName, numRequested, sink));
+				}
+			});
+
+		});
+	}
+
+	private abstract class PubSubPullTask implements Runnable {
+
+		protected final String subscriptionName;
+
+		protected final FluxSink<AcknowledgeablePubsubMessage> sink;
+
+		PubSubPullTask(String subscriptionName, FluxSink<AcknowledgeablePubsubMessage> sink) {
+			this.subscriptionName = subscriptionName;
+			this.sink = sink;
+		}
+
+		/**
+		 * Retrieve up to a specified number of messages, sending them to the subscription.
+		 * @param demand maximum number of messages to retrieve
+		 * @param block whether to wait for the first message to become available
+		 * @return number of messages retrieved
+		 */
+		protected int pullToSink(long demand, boolean block) {
+			int numMessagesToPull = demand > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) demand;
+
+			List<AcknowledgeablePubsubMessage> messages =  PubSubReactiveFactory.this.subscriberOperations.pull(
+					this.subscriptionName, numMessagesToPull, !block);
+
+			if (!this.sink.isCancelled()) {
+				messages.forEach(sink::next);
```

Comment:
I wonder what is wrong with sending the whole `List<AcknowledgeablePubsubMessage>` batch to the sink as a single event and then use a `flatMapIterable(Function.identity())` before returning the `Flux` to end-user...

Reply:
`flatMapIterable` changes demand, which could look unintuitive to the consumer (if subscription requests 3 messages, `flatMapIterable()` will cause 256 to be requested instead).

Also (this is not a good reason, just relevant) when I was testing, I ran into an issue with additional demand not being propagated from HTTP request to the Pub/Sub flux even though the initial request was fulfilled. I'd like to reproduce a minimal case to understand whether this behaves as intended or is a bug.

Reply:
OK. Then let's deffer it to @smaldini .

I'm not sure how to be with this situation then 😄

Reply:
Hi! Great work @elefeint! We were also thinking of implementing our own Reactor Pubsub client due to improper back-pressure semantics which is giving us quite a lot of headache and thankfully @bsideup pointed this ticket to us. About this specific `messages.forEach(sink::next)` line, I have a question: If the `demand` is a big number, wouldn't you be fetching a gigantic batch of messages from Pubsub and choking the consumer via over-production? I would rather cap it by `Math.min(demand, maxDemand)` (e.g., `maxDemand=100` by default) for each `request(demand)` and wait for any follow up `request()`s that the consumer will signal at its convenience. Otherwise my impression is that we will be periodically bombarding the consumer with messages regardless of its availability. Is there a detail I am missing?

Reply:
If the consumer is specifying the large demand, isn't it a signal that it can handle that number of messages? Certainly throttling can be implemented, but it may be unintuitive to consumers that do want a larger batch.

Reply:
Most of the time consumers do not specify anything and simple `subscribe()` just triggers a `request(Long.MAX_VALUE)`. Though in the case of a message queue poller, "*Apparently the consumer is asking for a cargo ship load of messages.*" does not sound like a practical assumption to me. Consider the following case:

1. There are 1e6 messages in the queue.
2. `request(unbounded)` triggers a continuous `messages.forEach(sink::next)` sequence.
3. Scheduler threads start working on the initial tasks and the rest quickly get piled up in the work queue of the `ScheduledExecutorService`.
4. After a while, majority of the messages timeout, work queue starts choking memory, and every task threads pick up turn out to be obsolete.

This is the problem I am dealing for the last couple of months at work. Let me know if there is a way I can work around this.

Reply:
@vy I understand now. As a practical matter, would adding[ limitRequest() ](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#limitRequest-long-) or [limitRate()](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#limitRate-int-) operators to the invocation chain work?

@smaldini The conversation about default behavior of subscribe() is an interesting one.
Would it make sense to open an issue on reactor-core? Unbounded demand is not the most common usecase, so perhaps a `.subscribeWithInitialDemand()` operator makes sense?

Reply:
@elefeint I suppose a combination of `limit{Request,Rate}()` would do the trick, though it still expects an explicit action from the consumer side. I would still favor shipping a solution with decent defaults minimizing any effort from user side.

I think `subscribeWithInitialDemand()` is sort of against the Reactor's fluent API encouraging composition. Isn't it just a shortcut for `.limitRequest(n).subscribe()`?

Nevertheless, @smaldini, we would really appreciate your insights on the general back-pressure issue. (@simonbasle, is this something you can help us out?)

Reply:
FYI `limitRate` only, not `limitRequest`. `limitRequest` will cut the requests after the threshold specified as an argument

Reply:
@vy Would you mind creating an issue in this project for the `maxMessagesPerPoll` property? I've been thinking about this some more, and for unbounded demand scenario it makes sense to provide a configurable limit, especially if it's combined with a configurable project-wide default for `pollingPeriodMs`. 
The firehose of messages would likely still be the default, but smoothing message arrival by getting N messages per poll seems valuable.